### PR TITLE
WIP: Actually try deprecating non-tuple nd-indices

### PR DESCRIPTION
**Pull request by seberg**
Just here for discussion. I am not sure it isn't too brutal. Also there is a test failure right now, which I think is probably another bug...

### CODE REVIEW

**Comment by seberg** [comment]
I have updated this (resolved merge conflict. Obviously it is not ready, since the 1.9. release notes are probably not the right place ;), but could discuss this again.

**Comment by seberg** [comment]
Updated this a bit, but getting some weirder errors, weeding out everything, especially in the masked array stuff seems a bit tricky. Anyone who wants to pick it up, be my guest, it is getting a bit annoying :).

File: doc/release/1.11.0-notes.rst [hunk 1]
```diff
@@ -39,6 +39,13 @@ DeprecationWarning to error
 FutureWarning to changed behavior 
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+* Multidimensional indexing with anything but a base class tuple is
+  deprecated. This means that code such as ``arr[[slice(None)]]`` has to
+  be changed to ``arr[tuple([slice(None)])]``. This is necessary to avoid
```
Comment (shoyer) [inline doc/release/1.11.0-notes.rst]: I would write: `arr[(slice(None),)]` or simply `arr[slice(None),]`

Comment (seberg) [inline doc/release/1.11.0-notes.rst]: Hehe, yeah, though that would somewhat defeat the purpose ;). should make the example longer... Anyway, doesn't look like 1.11 change.
Hmmm, reading my last comment, part of me wonders why the masked arrays are always so tricky when doing indexing changed.

File: numpy/core/src/multiarray/mapping.c [hunk 2]
```diff
@@ -221,18 +225,49 @@ prepare_index(PyArrayObject *self, PyObject *index,
...
+                        "using a non-tuple sequence for multidimensional "
+                        "indexing is deprecated; use `arr[tuple(seq)]` "
+                        "instead of `arr[seq]`. In the future this will be "
+                        "interpreted as an array index.") < 0) {
+                    return -1;
+                }
+            }
+            else if (make_tuple == 1) {
+                if (DEPRECATE_FUTUREWARNING(
+                        "multi dimensional indexing tuple was not base class. "
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c]: Is there a good reason to deprecate this? Allowing subclasses of tuple doesn't seem unreasonable, as there's still no ambiguity

Comment (seberg) [inline numpy/core/src/multiarray/mapping.c]: No, probably not, I probably wasn't sure and just thought I will go with the extreme version and make it less extreme from there ;).

**Comment by eric-wieser** [comment]
@seberg: Can you make this editable so that I can pick it up?

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -1242,11 +1242,11 @@ def gradient(f, *varargs, **kwargs):
...
         otype = f.dtype
 
-    # Convert datetime64 data into ints. Make dummy variable `y`
-    # that is a view of ints if the data is datetime64, otherwise
+    # Convert datetime64 data into timedelta. Make dummy variable `y`
+    # that is a view of timedelta if the data is datetime64, otherwise
     # just set y equal to the the array `f`.
     if f.dtype.char in ["M", "m"]:
-        y = f.view('int64')
+        y = f.view(otype)
```
Comment (eric-wieser) [inline numpy/lib/function_base.py]: What happened here?

Comment (seberg) [inline numpy/lib/function_base.py]: there is some weird problems in gradient that made it work only if it did the wrong, fancy, indexing. I guess this maybe somehow fixed/worked around it, it is possible that the commit message explains more, but likely not.

Comment (seberg) [inline numpy/lib/function_base.py]: Would not rule out that other datetime fixes in the meantime may have made this better, or that changing things at another place would be more correct.

**Comment by seberg** [comment]
Anyway, the problem with this is less the work to make it work, but more making the decision to do this very noisy deprecation which affects quite a lot of things.

**Comment by eric-wieser** [comment]
One thing we could do is merge all the fixes that make numpy only ever index with tuples internally.

**Comment by seberg** [comment]
Yeah, I guess so, though unless you add a compile time option to give the warning or so (and use it during testing), there will be no guarantee for it in the future and its useless. There is a bit of a downer, that manual tuple conversion is slower (probably because python has kwarg parsing overhead or so), but there is no point being annoyed by it I suppose.

**Comment by eric-wieser** [comment]
@seberg, I went ahead and rebased this at #9686

**Comment by charris** [comment]
Should we close #4434?

**Comment by seberg** [comment]
Sure, why not, could always reopen anyway if there were a reason for it.

**Pull request closed without merging**